Staging and Dev builds

martindsq opened the pull request:
I created two apps in appstoreconnect: Nos (staging) and Nos (dev).
And two schemes and environments in Xcode: Nos Staging and Nos Dev.
Each environment has its own Keychain group, so the login is not shared among them, and its recommended to maintain different npubs so to have throwaway accounts there.
This means that you can do four things normally:

- Debug the app in the old (current) way, this uses `com.verse.Nos` (so it overrides the TestFlight installation), but could be useful to debug something specific to the release build.
- Debug the app in the new (proposed) way using the scheme Nos Dev, this uses `com.verse.Nos-dev`, with a different icon and lets you have installed it in your device alongside the release build. Normally, this is what you would prefer to do most of the time as a developer.
- Release a staging app so Linda can test using the scheme Nos Staging. You can do that by running `fastlane staging`, but GitHub action will do it automatically each time something is merged to main.
- Release a production build to final users, this continues to use the current scheme (which was left untouched btw).

Staging uses the same secrets as production. At least for now, it is useful so that we are sure that everything works fine:

- In Sentry, staging builds (again, released each time something is merged to main, and meant to be UATested) can be filtered because the events uses the staging environment. Releases that use Nos Dev (you can manually execute fastlane dev, but I don't see what's the use of that) will appear under the debug environment. Debug builds don't appear because they don't have the sentry api key.
- In Posthog, all events from `com.verse.Nos-staging` are filtered out, so only events from the production build (`com.verse.Nos`) appear there.

As TestFlight doesn't allow reusing the build number (something totally comprehensible), build numbers in Nos, Nos (staging) and Nos (dev) are independent. If we want to have unique build numbers across the three apps, we can modify the Fastfile to query the latest build number in the three apps and use the higher one. I didn't do that, but let me know if you consider that to be important.

Finally, you can run `fastlane certs` to install the certificates. Nos and Nos Dev are debuggable in my machine, but let me know if the provisioning profile doesn't work for you right away.

joshuatbrown commented:
I don't know if it matters much, but I almost want to have something in the app that tells us what commit was built. Or maybe just a git tag along the lines of `staging-0.1-99`. Or maybe none of that matters too much. I guess I just want to make sure that if we go through UAT on a build and want to release a build from that specific commit, we have a way to know which one it is. Does that seem like something we need to be concerned about? Or is that already handled?

joshuatbrown commented on .github/workflows/testflight-deploy.yml:
Maybe this is not a big deal but I think I'd expect `testflight-deploy.yml` to be the release deployment and any others to be named `testflight-deploy-staging.yml` or whatever. Does that make sense?

martindsq replied:
It makes sense. I'll leave them as it is for now because the actions that appear under Actions tab here on GitHub are compiled from main and give me the ability to run one of them from another branch (but I can't if the yml has a different name because it doesn't appear in the list).

martindsq replied:
I implemented your suggestion

joshuatbrown commented on .gitignore:
```diff
@@ -5,7 +5,7 @@
 
 Nos/Assets/ProductionSecrets.xcconfig
-Nos/Assets/StagingSecrets.xcconfig
+Nos/Assets/DevSecrets.xcconfig
```
If I'm reading this right, ProductionSecrets will be production, of course, then DevSecrets will be development. But then Secrets will be staging -- is that right? I think it's be nice to keep ProductionSecrets and DevSecrets and then have StagingSecrets if I'm understanding this correctly.

martindsq replied:
I added StagingSecrets again. I guess I wanted to show that, in real life, secrets for production and staging are the same because at the moment we don't have anything we want to differentiate. But you're right, let's have different secrets so that we set the basis for future additions.

joshuatbrown commented on Nos.xcodeproj/xcshareddata/xcschemes/Nos Dev.xcscheme:
I always like to avoid spaces in file names -- maybe these aren't an issue but I recently read about [spaces in file names breaking SwiftUI previews](https://makerspace.social/@tritchey/112090374678123151). That, combined with bash scripts that seem to sometimes break because they don't properly handle file paths with spaces, makes me hesitant to ever include spaces. Would you mind removing them?

joshuatbrown commented on Nos/Assets/Assets.xcassets/AppIcon (dev).appiconset/App_store_1024_1x_dev.png:
Just pointing out the space in this one and the staging asset, too, to make sure we get them all.

joshuatbrown commented on fastlane/Fastfile:
```diff
@@ -25,8 +25,8 @@ platform :ios do
 
-  desc "Push a new development build to TestFlight"
-  lane :dev do
+  desc "Push a new build to TestFlight"
```
Maybe add something like this to provide a bit more detail?
```suggestion
  desc "Push a new build to TestFlight using the provided scheme and identifier"
```

I know there are other params in `options` but I'm not sure we need to specify all of them. Thoughts?

joshuatbrown commented on fastlane/README.md:
````diff
@@ -56,7 +56,23 @@ For _fastlane_ installation instructions, see [Installing _fastlane_](https://do
+[bundle exec] fastlane ios release
+```
+
+Push a new Nos build to TestFlight
````
```suggestion
Push a new Nos Release build to TestFlight
```

joshuatbrown reviewed (changes requested):
A few thoughts on naming and comments. Overall I like this a lot, and I'm super glad we're going to have the ability to easily install multiple builds at once.

martindsq commented:
> I don't know if it matters much, but I almost want to have something in the app that tells us what commit was built. Or maybe just a git tag along the lines of `staging-0.1-99`. Or maybe none of that matters too much. I guess I just want to make sure that if we go through UAT on a build and want to release a build from that specific commit, we have a way to know which one it is. Does that seem like something we need to be concerned about? Or is that already handled?

Yeah, let's tag the staging and dev builds! I changed the fastfile code accordingly.

martindsq commented:
@joshuatbrown it's ready to be reviewed again

mplorentz commented:
Can we get some icons from @Chardot for the new apps? Doesn't have to a part of this PR, and happy to slot them in myself when they are ready.

martindsq commented:
> Can we get some icons from @Chardot for the new apps? Doesn't have to a part of this PR, and happy to slot them in myself when they are ready.

Sure. @Chardot was very busy at that time working on other tickets that had more priority and that's why I went ahead with my own icons. Just be sure to have @setch-l notified that Sebastian would be working on that.

joshuatbrown commented:
I had issues when I tried to build NosDev locally for the simulator. Martin helped me to determine that it was due to automatic signing being turned off. We discussed in Slack and it seems best to keep "Automatically manage signing" turned on for Dev builds. This means we won't be able to produce Dev builds from GitHub actions, but we don't see much of a use case for being able to to Dev builds from GH actions anyway, so this shouldn't be a problem.

joshuatbrown reviewed (dismissed):
I found that I'm not able to build NosDev in Xcode 15.3 but everything builds just fine in Xcode 15.2, so we don't need to be concerned with that for now.

I would suggest one more minor change: the Product Name and Bundle Display Name of the apps should be consistent. Currently we have "Nos (dev)" and "Nos Staging" for Product Name but "Nos (dev)" and "Nos (staging)" for Bundle Display Name. I prefer the format of "Nos Staging", so I propose changing to that, "Nos Dev" for the other, and then we can eliminate all parentheses.

mplorentz commented:
I am not able to build the NosDev scheme and deploy it on my real phone. I think we want automatic signing configured for that case specifically? Otherwise we'll have to add all our device UUIDs to a text file and use `match` to regenerate the certificates whenever they change.

I'd also like to see NosDev as the default scheme when the repo is checked out.

martindsq commented:
@mplorentz

I could make this work in Xcode 15.3, which is the problem @joshuatbrown was having. swift-custom-dump (a swift package made by Apple) was causing problems when compiling, and I tracked it down to two package dependencies we are using that depend on it: swift-dependencies and swiftui-navigation (both made by the same developer). We were using very old versions because both progressed from 0.x to 1.x since the time we added them and it was just a matter of updating them to the latest version. After this, I was forced to also update our GitHub action yaml files to make the ci use the latest macOS and xcode version (this was a long process of trying a change in the ci and see if it failed until I had something because there was something that was causing problems when switching to macOS-14 and later I found out we had to add `-skipMacroValidation` to xcodebuild to make it work as suggested in https://stackoverflow.com/questions/77302264/use-skipmacrovalidation-for-xcodebuild-command-when-using-an-external-package

I also updated the Product Name and Bundle Display Name as suggested by Josh.

And turned on the automatic signing for NosDev (something we figured it out with Josh but didn't commit it. The tradeoff is that we can't submit NosDev builds from Github unless we create an additional configuration without that turned on, like NosDev Debug and NosDev Release. Even though I think that would be the most complete solution, with @joshuatbrown we agreed that for the moment, it would be over-engineering it because we actually don't need to send NosDev builds from the CI.

And finally, I couldn't set NosDev as the default scheme. Looking at the Internet, I found that it is a user-specific setting, not a shared one, so I think there is not much to do here more than a suggestion in the `CONTRIBUTING` file if you feel it is necessary. I don't know honestly what Xcode takes as a default, I tried reordering the schemes and that didn't change any shared file, just a user-specific file which is in line with what I found. I guess, and this is a real guess, that Xcode uses the target with the same name as the project as a default.

mplorentz reviewed (changes requested):
Thanks for fixing up the CI! 

I really do want to keep the project building successfully on a fresh checkout, without needing to change the scheme or anything. I decided to play around with this, and I think I was able to make NosDev the default scheme by renaming it to "Nos". I renamed the old "Nos" scheme "NosProduction". Maybe you could test on a fresh checkout and see if you get the same results. I know it would require more changes to CI to change the schemes around again, but if you're good with that you can find my changes pushed up to `change-default-scheme`.

mplorentz commented on .github/workflows/test.yml:
```diff
@@ -11,18 +11,20 @@ on:
         uses: actions/checkout@main
       - uses: nomasystems/action-xcode-select@v0.1
+        with:
+          xcode-version: '15.3'
```
The other nice part about using .xcode-version is that fastlane reads from it too, so it's a nice safeguard against a release build being built with an unexpected version of Xcode.

martindsq commented:
@mplorentz I didn't considered renaming Nos, I thought we didn't want to touch that. Yeah, that would work. I'll update the yaml files, I'm already an expert after working on this ticket.

mplorentz commented:
I resolved the conflicts because they were minor and I think they were holding up the CI maybe.

mplorentz reviewed (approved):
This looks ready to merge to me!
